### ISSUE TITLE
Wheel support for aarch64

### ISSUE THREAD

**Issue by odidev** [pyca/pynacl#601]
**Summary**
Installing pynacl on aarch64 via pip using command "pip3 install pynacl " tries to build wheel from source code

**Problem description**
pynacl don't have wheel for aarch64 on PyPI repository. So, while installing pynacl via pip on aarch64, pip builds wheel for same resulting in it takes more time to install pynacl. Making wheel available for aarch64 will benefit aarch64 users by minimizing pynacl installation time.

**Expected Output**
Pip should be able to download pynacl wheel from PyPI repository rather than building it from source code.

@pynacl-team, please let me know if I can help you building wheel/uploading to PyPI repository. I am curious to make pynacl wheel available for aarch64. It will be a great opportunity for me to work with you.

**Comment by brubbel** [comment]
+1 Same for armhf (raspberry pi). Latest versions of docker-compose have pynacl in the requirements.

**Comment by odidev** [comment]
Ping @reaperhulk, can you please let me know your interest in releasing AArch64 wheel?

**Comment by odidev** [comment]
Gentele Reminder !!!

**Comment by Lukejo2** [comment]
+1 This takes 45 minutes to build while cross-compiling to aarch64 from a gitlab runner.

**Comment by odidev** [comment]
Hi @reaperhulk 

Thanks for reviewing and merging the PR.
Please let me know if any help is required from our side to add wheel support for aarch64.

Thanks

**Comment by reaperhulk** [comment]
The wheel building is the harder problem here.

Right now we use GHA to build wheels, download them, and then upload them. We don't store API tokens capable of uploading wheels with GH since we'd like to limit scope of compromise (although this could potentially be changed if there was sufficient reason).

However, GHA doesn't have AArch64, so we can't build wheels there at all unless we do some kind of qemu hack. And we can't arbitrarily trigger TravisCI jobs like we can GH actions (nor does it have storage -- all Travis style wheel generation needs to build and upload in one step), so it would be difficult to use Travis.

**Comment by Dish1306** [comment]
Hi @reaperhulk ,

We will look how qemu hack can be used on GHA for AArch64 and update you the steps, 
Will that be helpful ??

Thanks

**Comment by odidev** [comment]
Hi @reaperhulk ,

We will look how qemu hack can be used on GHA for aatch64 and update you the steps, 
Will that be helpful?

Thanks

**Comment by FlyingThunder** [comment]
Hello, i seem(?) to have a similar issue, installing via pip gives me the error "Could not build wheels for PyNaCl which use PEP 517 and cannot be installed directly" - however i am on Windows 10? I already tried it with --no-binary :all: which also didnt work. I need PyNaCl but i dont have much knowledge when it comes to pip and wheels so i have no idea how to solve this...

Is this a PyNaCl issue or related to pip, and if so, should i open a new issue for it?

**Comment by ianw** [comment]
(Moved from #612 as this is more a discussion)

> This approach is too slow unfortunately. If we used this there would be a 40+ minute window where, after a release, a new sdist was available but no wheels would be. We do have access to graviton2 now that we've migrated our org, but we'll have to change our wheel building approach to happen automatically on tag (and sort out how we want to handle pypi tokens) to switch to that model.

I will be happy to work the existing OpenDev Zuul wheel builds into this repository as done for pyca/cryptography, if that would be OK.  We can share most of the roles and make the jobs generic.  I think the Zuul bot would just need to be enabled here too.

Zuul could definitely upload sdists and linux wheels automatically on tags.  Unfortunately we don't have anywhere to build osx and windows though.  So I understand you want to keep everything together, but the offer is there :) 

> and sort out how we want to handle pypi tokens

FYI Zuul handles secrets for situations like this with the idea of post-review pipelines (https://zuul-ci.org/docs/zuul/reference/secret_def.html).  Zuul exports a key for each repo which you can encode a secret with (https://zuul-ci.org/docs/zuul/discussion/encryption.html); i.e. you keep the secret encrypted in your public job definitions and Zuul can decrypt it as it needs to.  You can only use that secret in trusted places -- i.e. the post-review release pipeline, that happens in response to tag events.  If a change has been tagged, then it's assumed only an authorized person has done that, and the tagged code has been reviewed and committed so is not doing anything silly like saying "hey, print out the value of this secret for me".  You obviously can not submit a pull request (that executes speculatively) that dumps a secret as that isn't a trusted context.

So in short, if the project added the openstackci pypi user, I could encrypt the secret I know for that, and we'd have a release pipeline job that would uploading using it.

**Comment by jryberg** [comment]
Hi,

Any progress?

**Comment by MartensCedric** [comment]
bump

**Comment by fuomag9** [comment]
Any progress on this? docker-compose requires this library and I can't install on lower-powered cpus without taking ages (or having dependency issues)

**Comment by ianw** [comment]
I'm still willing to set this up in opendev, if the bot is added here.  We now have resources from OSUOSL as well, making it even more reliable

**Comment by Jeremy-boo** [comment]
Is there any progress on this matter? I ran into the same problem using pip install google-apitools==0.5.31.

os: Linux harbor-arm-ci-04 5.4.0-40-generic #44-Ubuntu SMP Mon Jun 22 23:59:48 UTC 2020 aarch64 aarch64 aarch64 GNU/Linux

Dockerfile:
# photon based image
FROM photon:4.0
ENV LANG C.UTF-8

COPY --from=tool_builder /tool/tools.tar.gz /usr/local/bin

RUN tdnf install -y \
    wget \
    git \
    openjdk8 \
    iproute2 \
    iptables \
    build-essential \
    sed \
    openssl-devel \
    tar \
    which \
    procps-ng \
    gzip && \
    tdnf erase -y toybox && \
    tdnf install -y python3 python3-pip python3-setuptools httpd && \
    pip3 install --upgrade pip pyasn1 google-apitools==0.5.31 gsutil \
    robotframework==3.2.1 robotframework-sshlibrary robotframework-httplibrary \
    requests dbbot robotframework-seleniumlibrary==4.3.0 robotframework-pabot \
    robotframework-JSONLibrary hurry.filesize --upgrade && \
    tdnf clean all


error info:

  subprocess.CalledProcessError: Command '['make', 'check']' returned non-zero exit status 2.
  ----------------------------------------
  ERROR: Failed building wheel for pynacl
Failed to build pynacl
ERROR: Could not build wheels for pynacl which use PEP 517 and cannot be installed directly
The command '/bin/sh -c tdnf install -y     wget     git     openjdk8     iproute2     iptables     build-essential     sed     openssl-devel     tar     which     procps-ng     gzip &&     tdnf erase -y toybox &&     tdnf install -y python3 python3-pip python3-setuptools httpd &&     pip3 install --upgrade pip pyasn1 google-apitools==0.5.31 gsutil     robotframework==3.2.1 robotframework-sshlibrary robotframework-httplibrary     requests dbbot robotframework-seleniumlibrary==4.3.0 robotframework-pabot     robotframework-JSONLibrary hurry.filesize --upgrade &&     tdnf clean all' returned a non-zero code: 1

**Comment by reaperhulk** [comment]
We now ship arm wheels!
